04-Parallel-Simulations-With-MPI/04-Running-Multiple-Simulations-With-Partitions.ipynb

```diff
@@ -0,0 +1,235 @@
+    "\n",
+    "fn = os.path.join(os.getcwd(), 'trajectory0.gsd')\n",
+    "![ -e \"$fn\" ] && rm \"$fn\"\n",
+    "fn = os.path.join(os.getcwd(), 'trajectory1.gsd')\n",
+    "![ -e \"$fn\" ] && rm \"$fn\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Partitioning communicators\n",
+    "\n",
+    "So far in this tutorial you have seen how executing `mpirun -n 4 python3 script.py` will use domain decomposition to run *1* simulation split across 4 ranks via **domain decomposition**.\n",
+    "What if you wanted to run *2 different* simulations, each on *2* **ranks** with this command?\n",
+    "Or *4 different* simulations each on *1* **rank**?\n",
+    "This is called **partitioning** the MPI communicator.\n",
+    "\n",
+    "In HOOMD-blue, you can do this by defining a non-default Communicator and specifying the ``ranks_per_partition`` argument.\n",
+    "Then you can use the ``communicator.partition`` as an identifier in your script to change input parameters."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "\u001b[0;32mimport\u001b[0m \u001b[0mhoomd\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0mcommunicator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhoomd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCommunicator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mranks_per_partition\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'Hello from partition {communicator.partition} rank {communicator.rank}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "%pycat hello_partition.py"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Hello from partition 0 rank 1\n",
+      "Hello from partition 1 rank 0\n",
+      "Hello from partition 1 rank 1\n",
+      "Hello from partition 0 rank 0\n"
+     ]
+    }
+   ],
+   "source": [
+    "!mpirun -n 4 python3 hello_partition.py"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In partitioned simulations, all ranks *within a given partition* must have the same input file, operations, and parameters.\n",
+    "However, the input file, operations and/or parameters can differ between partitions.\n",
+    "Use this to run many simulations with different temperatures, different initial configurations, or as will be shown in this example, different random number seeds.\n",
```

Comment:
In the example below, we use `sim.seed = communicator.partition` to generate seeds for each partition. If I want to set different temperatures, the only way I can think of is to do something like 
```python
for temp in temp_list:
    if communicator.partition == partition_number: 
        langevin = hoomd.md.methods.Langevin(kT=temp, filter=hoomd.filter.All())
```
Is there a better/ more efficient way to do this?

Reply:
Yes, there is a better pattern:
```python
langevin = hoomd.md.methods.Langevin(kT=temp_list[communicator.partition], filter=hoomd.filter.All())
```
I will include this in the example.